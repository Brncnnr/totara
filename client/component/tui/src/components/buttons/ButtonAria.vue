<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2022 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Simon Chester <simon.chester@totaralearning.com>
  @module tui
-->

<script>
import {
  cloneVNode,
  mergeListeners,
  normalizeListenerValue,
} from '../../js/internal/vnode';

const KEY_ENTER = 13;
const KEY_SPACE = 32;

export default {
  methods: {
    getDisabled() {
      return this.$el.getAttribute('disabled');
    },
    handleClick(event) {
      if (this.getDisabled()) {
        event.preventDefault();
        // Replicate native behavior: no click events on disabled buttons
        // (including bubbling)
        event.stopPropagation();
      }
    },
    handleKeyDown(event) {
      if (event.keyCode === KEY_SPACE) {
        event.preventDefault();
      } else if (event.keyCode === KEY_ENTER) {
        event.preventDefault();
        this.$el.click();
      }
    },
    handleKeyUp(event) {
      if (event.keyCode === KEY_SPACE) {
        event.preventDefault();
        this.$el.click();
      }
    },
    wrapClickListener(fn) {
      return event => {
        if (!this.getDisabled()) {
          fn(event);
        }
      };
    },
  },

  render() {
    const content = this.$scopedSlots.default && this.$scopedSlots.default();
    if (!content || content.length == 0) {
      return null;
    }
    if (content.length > 1) {
      console.error('ButtonAria may only contain a single child');
      return null;
    }

    let vnode = content[0];

    if (vnode.componentOptions) {
      console.error(
        'The content of ButtonAria must be a plain HTML element, not a component'
      );
      return null;
    }

    vnode = cloneVNode(vnode);

    const isDisabled = vnode.data.attrs && Boolean(vnode.data.attrs.disabled);

    vnode.data.attrs = {
      role: 'button',
      tabindex: isDisabled ? null : '0',
      'aria-disabled': isDisabled ? 'true' : null,
      ...vnode.data.attrs,
    };

    const onListeners = { ...vnode.data.on };

    // Replicate native behavior by suppressing click handlers
    const clickListener =
      onListeners.click && normalizeListenerValue(onListeners.click);
    if (clickListener) {
      onListeners.click = clickListener.map(this.wrapClickListener);
    }

    vnode.data.on = mergeListeners(
      {
        click: this.handleClick,
        keydown: this.handleKeyDown,
        keyup: this.handleKeyUp,
      },
      onListeners
    );

    return vnode;
  },
};
</script>
