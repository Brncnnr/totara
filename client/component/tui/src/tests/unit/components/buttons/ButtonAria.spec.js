/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2022 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import { mount } from '@vue/test-utils';
import ButtonAria from 'tui/components/buttons/ButtonAria';
import { axe } from 'jest-axe';

// Simple button implementation for testing
const SimpleButton = {
  props: ['disabled'],
  render(h) {
    return h(ButtonAria, [
      h(
        'div',
        {
          attrs: { disabled: this.disabled },
          on: { click: e => this.$emit('click', e) },
        },
        ['foo']
      ),
    ]);
  },
};

describe('ButtonAria', () => {
  it('has accessible attributes', () => {
    const wrapper = mount(SimpleButton);
    expect(wrapper.attributes()).toEqual({
      role: 'button',
      tabindex: '0',
    });
  });

  it('emits click event on button click', () => {
    const click = jest.fn();
    const wrapper = mount(SimpleButton, {
      listeners: { click },
    });

    wrapper.trigger('click');
    expect(click).toHaveBeenCalled();
    click.mockClear();

    wrapper.trigger('keydown.enter');
    expect(click).toHaveBeenCalled();
    click.mockClear();

    wrapper.trigger('keyup.space');
    expect(click).toHaveBeenCalled();
    click.mockClear();
  });

  it('behaves appropriately when button is disabled', () => {
    const click = jest.fn();
    const wrapper = mount(SimpleButton, {
      propsData: { disabled: true },
      listeners: { click },
    });
    wrapper.trigger('click');
    expect(click).not.toHaveBeenCalled();
    expect(wrapper.attributes()).toEqual({
      role: 'button',
      disabled: 'disabled',
      'aria-disabled': 'true',
    });
  });

  it('matches snapshot', () => {
    const wrapper = mount(SimpleButton);
    expect(wrapper.element).toMatchSnapshot();

    const disabledWrapper = mount(SimpleButton, {
      propsData: { disabled: true },
    });
    expect(disabledWrapper.element).toMatchSnapshot('disabled');
  });

  it('should not have any accessibility violations', async () => {
    const wrapper = mount(SimpleButton);
    const results = await axe(wrapper.element);
    expect(results).toHaveNoViolations();
  });
});
