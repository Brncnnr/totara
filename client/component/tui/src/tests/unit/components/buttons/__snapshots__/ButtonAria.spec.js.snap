// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonAria matches snapshot 1`] = `
<div
  role="button"
  tabindex="0"
>
  foo
</div>
`;

exports[`ButtonAria matches snapshot: disabled 1`] = `
<div
  aria-disabled="true"
  disabled="disabled"
  role="button"
>
  foo
</div>
`;
